program/src/guards/freeze_token_payment.rs

```diff
@@ -0,0 +1,343 @@
+            //   1. `[signer]` Candy Guard authority.
+            //   2. `[]` System program account.
+            //   3. `[writable]` Associate token account of the Freeze PDA (seeds `[freeze PDA pubkey, nft mint pubkey]`).
+            //   4. `[]` Token mint account.
+            FreezeInstruction::Initialize => {
+                msg!("Instruction: Initialize (FreezeTokenPayment guard)");
+
+                if route_context.candy_guard.is_none() || route_context.candy_machine.is_none() {
+                    return err!(CandyGuardError::Uninitialized);
+                }
+
+                if route_context.candy_guard.is_none() || route_context.candy_machine.is_none() {
+                    return err!(CandyGuardError::Uninitialized);
+                }
+
+                let destination = if let Some(guard_set) = &route_context.guard_set {
+                    if let Some(freeze_guard) = &guard_set.freeze_token_payment {
+                        freeze_guard.destination_ata
+                    } else {
+                        return err!(CandyGuardError::FreezeGuardNotEnabled);
+                    }
+                } else {
+                    return err!(CandyGuardError::FreezeGuardNotEnabled);
+                };
+
+                // initializes the freeze pda (the check of the authority as signer is done
+                // during the initialization)
+                initialize_freeze(ctx, route_context, data, destination)?;
+
+                // initializes the freeze ata
+
+                let freeze_pda = get_account_info(ctx, 0)?;
+                let system_program = get_account_info(ctx, 2)?;
+                let freeze_ata = get_account_info(ctx, 3)?;
+                let token_mint = get_account_info(ctx, 4)?;
+
+                assert_keys_equal(
+                    &get_associated_token_address(freeze_pda.key, token_mint.key),
+                    freeze_ata.key,
+                )?;
+
+                invoke(
+                    &create_associated_token_account(
+                        ctx.accounts.payer.key,
+                        freeze_pda.key,
+                        token_mint.key,
+                    ),
+                    &[
+                        ctx.accounts.payer.to_account_info(),
+                        freeze_ata.to_account_info(),
+                        freeze_pda.to_account_info(),
+                        token_mint.to_account_info(),
+                        system_program.to_account_info(),
+                    ],
+                )?;
+
+                Ok(())
+            }
+            FreezeInstruction::Thaw => {
+                msg!("Instruction: Thaw (FreezeTokenPayment guard)");
+                thaw_nft(ctx, route_context, data)
+            }
+
+            FreezeInstruction::UnlockFunds => {
+                msg!("Instruction: Unlock Funds (FreezeTokenPayment guard)");
+                unlock_funds(ctx, route_context)
+            }
```

Comment:
Missing the list of required accounts comments. I think it should be listed here even though its already stated above the `unlock_funds` method and in teh `thaw_nft` ix in the sol payment

Reply:
Added comments.